// redis-om/dist/entity/fields/entity-string-array-field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const entity_field_1 = tslib_1.__importDefault(require("./entity-field"));
class EntityStringArrayField extends entity_field_1.default {
    toRedisHash() {
        const data = {};
        if (this.value !== null)
            data[this.name] = this.value.join(this.separator);
        return data;
    }
    fromRedisHash(value) {
        this.value = value.split(this.separator);
    }
    validateValue(value) {
        super.validateValue(value);
        if (value !== null && !this.isArray(value))
            throw Error(`Expected value with type of 'string[]' but received '${value}'.`);
    }
    convertValue(value) {
        if (this.isArray(value)) {
            return value.map((v) => v.toString());
        }
        return super.convertValue(value);
    }
    get separator() {
        return this.fieldDef.separator ?? '|';
    }
    isArray(value) {
        return Array.isArray(value);
    }
}
exports.default = EntityStringArrayField;
