// redis-om/dist/schema/builders/schema-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class SchemaBuilder {
    schema;
    constructor(schema) {
        this.schema = schema;
    }
    get redisSchema() {
        const redisSchema = [];
        Object.keys(this.schema.definition).forEach(field => {
            redisSchema.push(...this.buildEntry(field));
        });
        return redisSchema;
    }
    buildCaseInsensitive(field) {
        return field.caseSensitive ? ['CASESENSITIVE'] : [];
    }
    buildIndexed(field) {
        return field.indexed ?? this.schema.indexedDefault ? [] : ['NOINDEX'];
    }
    buildStemming(field) {
        return field.stemming ?? true ? [] : ['NOSTEM'];
    }
    buildPhonetic(field) {
        return field.matcher ? ['PHONETIC', field.matcher] : [];
    }
    buildSeparable(field) {
        return ['SEPARATOR', field.separator || '|'];
    }
    buildSortable(field) {
        return field.sortable ? ['SORTABLE'] : [];
    }
    buildNormalized(field) {
        return field.normalized ?? true ? [] : ['UNF'];
    }
    buildWeight(field) {
        return field.weight ? ['WEIGHT', field.weight.toString()] : [];
    }
}
exports.default = SchemaBuilder;
