// redis-om/dist/search/results-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonSearchResultsConverter = exports.HashSearchResultsConverter = exports.SearchResultsConverter = void 0;
class SearchResultsConverter {
    results;
    schema;
    constructor(schema, results) {
        this.schema = schema;
        this.results = results;
    }
    get count() {
        const [count] = this.results;
        return Number.parseInt(count);
    }
    get ids() {
        return this.keys.map(key => key.replace(/^.*:/, ""));
    }
    get keys() {
        const [_count, ...keysAndValues] = this.results;
        return keysAndValues.filter((_entry, index) => index % 2 === 0);
    }
    get values() {
        const [_count, ...keysAndValues] = this.results;
        return keysAndValues.filter((_entry, index) => index % 2 !== 0);
    }
    get entities() {
        const ids = this.ids;
        const values = this.values;
        return values.map((array, index) => this.arrayToEntity(ids[index], array));
    }
}
exports.SearchResultsConverter = SearchResultsConverter;
class HashSearchResultsConverter extends SearchResultsConverter {
    arrayToEntity(id, array) {
        const keys = array.filter((_entry, index) => index % 2 === 0);
        const values = array.filter((_entry, index) => index % 2 !== 0);
        const hashData = keys.reduce((object, key, index) => {
            object[key] = values[index];
            return object;
        }, {});
        const entity = new this.schema.entityCtor(this.schema, id);
        entity.fromRedisHash(hashData);
        return entity;
    }
}
exports.HashSearchResultsConverter = HashSearchResultsConverter;
class JsonSearchResultsConverter extends SearchResultsConverter {
    arrayToEntity(id, array) {
        const index = array.findIndex(value => value === '$') + 1;
        const jsonString = array[index];
        const jsonData = JSON.parse(jsonString);
        const entity = new this.schema.entityCtor(this.schema, id);
        entity.fromRedisJson(jsonData);
        return entity;
    }
}
exports.JsonSearchResultsConverter = JsonSearchResultsConverter;
